### PR TITLE
The `AbstractTBTCDepositor` contract

### PR DESCRIPTION
**Pull request by lukasz-zimnoch**
Refs: https://github.com/keep-network/tbtc-v2/issues/750

Here we introduce the `AbstractTBTCDepositor ` contract meant to facilitate integration of protocols aiming to use tBTC as an underlying Bitcoin bridge. 

### Usage of the `AbstractTBTCDepositor` contract

An integrator willing to leverage the `AbstractTBTCDepositor` contract is supposed to:
- Create a child contract inheriting from this abstract contract
- Call the `__AbstractTBTCDepositor_initialize` initializer function
- Use the `_initializeDeposit` and `_finalizeDeposit` as part of their business logic in order to initialize and finalize deposits.

Example code:

```solidity                                                           
// Example upgradeable integrator contract.                   
contract ExampleTBTCIntegrator is AbstractTBTCDepositor, Initializable {          
    /// @custom:oz-upgrades-unsafe-allow constructor                           
    constructor() {                                                            
        // Prevents the contract from being initialized again.                 
        _disableInitializers();                                                
    }                                                                          
                                                                               
    function initialize(                                                       
        address _bridge,                                                       
        address _tbtcVault                                                     
    ) external initializer {                                                   
        __AbstractTBTCDepositor_initialize(_bridge, _tbtcVault);                  
    }                                                                          
                                                                               
    function startProcess(                                                     
        IBridgeTypes.BitcoinTxInfo calldata fundingTx,                         
        IBridgeTypes.DepositRevealInfo calldata reveal                         
    ) external {                                                               
        // Embed necessary context as extra data.                              
        bytes32 extraData = ...;                                               
                                                                               
        uint256 depositKey = _initializeDeposit(                               
            fundingTx,                                                         
            reveal,                                                            
            extraData                                                          
        );                                                                     
                                                                               
        // Use the depositKey to track the process.                            
    }                                                                          
                                                                               
    function finalizeProcess(uint256 depositKey) external {                    
        (uint256 tbtcAmount, bytes32 extraData) = _finalizeDeposit(depositKey);
                                                                               
        // Do something with the minted TBTC using context                     
        // embedded in the extraData.                                          
    }                                                                          
}                                                                                                                                           
```

### The `tbtcAmount` caveat

The `tbtcAmount` computed by `_calculateTbtcAmount` and returned by `_finalizeDeposit` may not correspond to the actual amount of TBTC minted for the deposit. Although the treasury fee cut upon minting is known precisely, this is not the case for the optimistic minting fee and the Bitcoin transaction fee. To overcome that problem, this contract just takes the current maximum values of both fees, at the moment of deposit finalization. For the great majority of the deposits, such an algorithm will return a `tbtcAmount` slightly lesser than the actual amount of TBTC minted for the deposit. This will cause some TBTC to be left in the contract and ensure there is enough liquidity to finalize the deposit. However, in some rare cases, where the actual values of those fees change between the deposit minting and finalization, the `tbtcAmount` returned by this function may be greater than the actual amount of TBTC minted for the deposit. If this happens and the reserve coming from previous deposits leftovers does not provide enough liquidity, the deposit will have to wait for finalization until the reserve is refilled by subsequent deposits or a manual top-up. The integrator is responsible for handling such cases.

### CODE REVIEW

File: solidity/contracts/bridge/DepositorProxy.sol [hunk 1]
```diff
@@ -0,0 +1,151 @@
...
+// TODO: Move to another directory?
+abstract contract DepositorProxy {
+    using BTCUtils for bytes;
+
+    /// @notice Multiplier to convert satoshi to TBTC token units.
+    uint256 public constant SATOSHI_MULTIPLIER = 10**10;
+
+    Bridge public bridge;
+    TBTCVault public tbtcVault;
+    TBTC public tbtcToken;
```
Comment (nkuba) [inline solidity/contracts/bridge/DepositorProxy.sol:38]: Do we need it? It doesn't seem to be used.

File: solidity/contracts/bridge/DepositorProxy.sol [hunk 2]
```diff
@@ -0,0 +1,151 @@
...
+        bridge = Bridge(_bridge);
+        tbtcVault = TBTCVault(_tbtcVault);
+        tbtcToken = TBTC(_tbtcToken);
+    }
+
+    function initializeDeposit(
+        BitcoinTx.Info calldata fundingTx,
+        Deposit.DepositRevealInfo calldata reveal,
+        bytes32 extraData
+    ) internal {
```
Comment (nkuba) [inline solidity/contracts/bridge/DepositorProxy.sol:63]: Can this function return the calculated `depositKey`?

File: solidity/contracts/bridge/DepositorProxy.sol (hunk 1 above)
Comment (lukasz-zimnoch) [reply solidity/contracts/bridge/DepositorProxy.sol]: Yeah, we don't. I was initially thinking about a way to manage the TBTC balance from the abstract contract level but it seems better to push that responsibility to the child contracts. See 8c068d4f

File: solidity/contracts/bridge/DepositorProxy.sol (hunk 2 above)
Comment (lukasz-zimnoch) [reply solidity/contracts/bridge/DepositorProxy.sol]: Of course! See fb2acde8

File: solidity/contracts/bridge/DepositorProxy.sol [hunk 3]
```diff
@@ -0,0 +1,151 @@
...
+import {BTCUtils} from "@keep-network/bitcoin-spv-sol/contracts/BTCUtils.sol";
+
+import "./BitcoinTx.sol";
+import "./Bridge.sol";
+import "./Deposit.sol";
+import "./../vault/TBTCVault.sol";
+import "./../vault/TBTCOptimisticMinting.sol";
+import "./../token/TBTC.sol";
+
+// TODO: Make it safe for upgradeable contracts.
```
[the remark this answers is not in the document]
Comment (lukasz-zimnoch) [reply solidity/contracts/bridge/DepositorProxy.sol]: See https://github.com/keep-network/tbtc-v2/pull/778/commits/7dc4048a0b91a9523690d23574a6fa3838d5f540

File: solidity/contracts/integrator/TBTCDepositorProxy.sol [hunk 4]
```diff
@@ -0,0 +1,358 @@
...
+// ██████████████     ▐████▌     ██████████████
+// ██████████████     ▐████▌     ██████████████
+//               ▐████▌    ▐████▌
+//               ▐████▌    ▐████▌
+//               ▐████▌    ▐████▌
+//               ▐████▌    ▐████▌
+//               ▐████▌    ▐████▌
+//               ▐████▌    ▐████▌
+
+pragma solidity 0.8.17;
```
Comment (nkuba) [inline solidity/contracts/integrator/TBTCDepositorProxy.sol:16]: As this is an abstract contract that will be implemented in other projects, we should give more flexibility around pragma to the integrator.

Comment (lukasz-zimnoch) [reply solidity/contracts/integrator/TBTCDepositorProxy.sol]: Good point. See b7ed96ea

File: solidity/contracts/integrator/TBTCDepositorProxy.sol [hunk 5]
```diff
@@ -0,0 +1,358 @@
...
+
+import {BTCUtils} from "@keep-network/bitcoin-spv-sol/contracts/BTCUtils.sol";
+import {Initializable} from "@openzeppelin/contracts-upgradeable/proxy/utils/Initializable.sol";
```
[the remark this answers is not in the document]
Comment (lukasz-zimnoch) [reply solidity/contracts/integrator/TBTCDepositorProxy.sol:19]: I think we can do it just for this abstract contract. ERC7201 seems to be worth it. Let me see.

[the remark this answers is not in the document]
Comment (lukasz-zimnoch) [reply solidity/contracts/integrator/TBTCDepositorProxy.sol]: Unfortunately, this won't be so simple. Migration from OZ v4 to v5 is non-trivial and will incur the upgrade of the OZ Upgrade plugin. This may break things while dealing with existing tBTC contract deployments. I don't want to take this risk right now.

However, to make `TBTCDepositorProxy` usable with both OZ v4 and v5, I decided to remove the dependency on `Initializable` in favor of a custom guard in the `__TBTCDepositorProxy_initialize` initializer. Regarding the storage layout protection, I think we should keep using storage gaps in this case. That gives the best flexibility as child contracts of `TBTCDepositorProxy` can still inherit from OZ v5. This is safe because ERC-7201 namespaces used by OZ v5 are safe against collisions with the standard storage layout used by Solidity (source: [ERC-7201 abstract](https://eips.ethereum.org/EIPS/eip-7201#abstract)). I have also explored the possibility of using ERC-7201 in `TBTCDepositorProxy`. Unfortunately, this is not an option because the OZ Upgrade Plugin raises an error for Solidity versions `<0.8.20` ([source](https://docs.openzeppelin.com/upgrades-plugins/1.x/writing-upgradeable#namespaced-storage-layout)) and `TBTCDepositorProxy` must support all `0.8.x` versions.

File: solidity/test/integrator/TBTCDepositorProxy.test.ts [hunk 6]
```diff
@@ -0,0 +1,317 @@
...
+            fixture.reveal,
+            fixture.extraData
+          )
+        })
+
+        after(async () => {
+          await restoreSnapshot()
+        })
+
+        it("should reveal the deposit to the Bridge", async () => {
```
Comment (nkuba) [inline solidity/test/integrator/TBTCDepositorProxy.test.ts:128]: A test to check that the deposit key is returned by `initializeDeposit` is missing.

File: solidity/test/integrator/TBTCDepositorProxy.test.ts [hunk 7]
```diff
@@ -0,0 +1,317 @@
...
+              .withArgs(
+                fixture.expectedDepositKey,
+                expectedTbtcAmount,
+                fixture.extraData
+              )
+          })
+        })
+      })
+    })
+  })
```
Comment (nkuba) [inline solidity/test/integrator/TBTCDepositorProxy.test.ts:316]: Would you be interested in adding tests for `calculateDepositKey` function to cover a scenario where the fees are set to zero?

File: solidity/contracts/integrator/TBTCDepositorProxy.sol [hunk 8]
```diff
@@ -0,0 +1,331 @@
...
+
+        // slither-disable-next-line reentrancy-events
+        emit DepositFinalized(
+            depositKey,
+            tbtcAmount,
+            /* solhint-disable-next-line not-rely-on-time */
+            uint32(block.timestamp)
+        );
+
+        onDepositFinalized(depositKey, tbtcAmount, deposit.extraData);
```
Comment (nkuba) [inline solidity/contracts/integrator/TBTCDepositorProxy.sol:222]: 🤔 Since the `finalizeDeposit` function is internal and expected to be overridden in the implementation contract, would it make more sense to return the values instead of passing them in a callback?
With the current approach, we expect the contract to implement both `finalizeDeposit` and `onDepositFinalized`, and they are invoked together.

What if we rename the internal function to `_finalizeDeposit` and return the values:
```suggestion
        return (tbtcAmount, deposit.extraData);
```

the implementation could look like this:
```solidity
    function finalizeDeposit(uint256 depositKey) public override {
        uint256 tbtcAmount, bytes32 extraData = _finalizeDeposit(depositKey);

        (address receiver, uint16 referral) = decodeExtraData(extraData);

        request.amountToStake = tbtcAmount;
        request.receiver = receiver;
        request.referral = referral;
    }
```

Comment (lukasz-zimnoch) [reply solidity/contracts/integrator/TBTCDepositorProxy.sol:222]: Actually `finalizeDeposit` is not expected to be overridden (it's not `virtual`) but rather called from a child's external function. I was also thinking about returning the parameters but the solution with `onDepositFinalized` seemed more clear and intuitive to me. I don't have a strong preference though. If you find return params more convenient, I can go that way.

File: solidity/test/integrator/TBTCDepositorProxy.test.ts (hunk 6 above)
Comment (lukasz-zimnoch) [reply solidity/test/integrator/TBTCDepositorProxy.test.ts:128]: Addressed as part of a31abf46

File: solidity/test/integrator/TBTCDepositorProxy.test.ts (hunk 7 above)
Comment (lukasz-zimnoch) [reply solidity/test/integrator/TBTCDepositorProxy.test.ts:314]: Yeah, it's good to cover those paths just in case. See 9efd0d0c

File: solidity/contracts/test/TestTBTCDepositorProxy.sol [hunk 9]
```diff
@@ -0,0 +1,171 @@
...
+        _deposits[depositKey] = request;
+
+        emit DepositRevealed(depositKey);
+    }
+
+    function sweepDeposit(uint256 depositKey) external {
+        require(_deposits[depositKey].revealedAt != 0, "Deposit not revealed");
+        require(_deposits[depositKey].sweptAt == 0, "Deposit already swept");
+        /* solhint-disable-next-line not-rely-on-time */
+        _deposits[depositKey].sweptAt = uint32(block.timestamp);
```
Comment (nkuba) [inline solidity/contracts/test/TestTBTCDepositorProxy.sol:96]: In the integrator tests, we will need tokens to be minted to the depositor. I'm wondering whether we could mint them here in `sweepDeposit` and `finalizeOptimisticMintingRequest` or leave it to the integrator.

If leaving it to the integrator it would be worth making `sweepDeposit` and `finalizeOptimisticMintingRequest` public virtual, so the integrator can reuse the test contracts from this file and just extend it with minting.

File: solidity/contracts/integrator/TBTCDepositorProxy.sol [hunk 10]
```diff
@@ -0,0 +1,309 @@
...
+    ///        (in the context of this contract) yet.
+    ///      - The deposit must be finalized on the Bridge side. That means the
+    ///        deposit must be either swept or optimistically minted.
+    /// @dev IMPORTANT NOTE: The tbtcAmount returned by this function is an
+    ///      approximation. See documentation of the `calculateTbtcAmount`
+    ///      responsible for calculating this value for more details.
+    // slither-disable-next-line dead-code
+    function _finalizeDeposit(uint256 depositKey)
+        internal
+        returns (uint256 tbtcAmount, bytes32 extraData)
```
Comment (nkuba) [inline solidity/contracts/integrator/TBTCDepositorProxy.sol:183]: Let's say the integrator wants to charge further fees on the deposit, what do you think about additionally returning the gross initial deposit amount here (multiplied by the satoshi multiplier)?

File: solidity/contracts/test/TestTBTCDepositorProxy.sol [hunk 11]
```diff
@@ -0,0 +1,171 @@
...
+
+        require(
+            _deposits[depositKey].revealedAt == 0,
+            "Deposit already revealed"
+        );
+
+        IBridgeTypes.DepositRequest memory request;
+
+        request.depositor = msg.sender;
+        request.amount = uint64(10 * 1e8); // 10 BTC
```
Comment (nkuba) [inline solidity/contracts/test/TestTBTCDepositorProxy.sol:79]: Can we set it to the actual funding deposit amount?

```solidity
        bytes memory fundingOutput = fundingTx
            .outputVector
            .extractOutputAtIndex(reveal.fundingOutputIndex);

        uint64 fundingOutputAmount = fundingOutput.extractValue();

        request.amount = fundingOutputAmount;
```

File: solidity/contracts/test/TestTBTCDepositorProxy.sol [hunk 12]
```diff
@@ -0,0 +1,171 @@
...
+        /* solhint-disable-next-line not-rely-on-time */
+        request.revealedAt = uint32(block.timestamp);
+        request.vault = reveal.vault;
+        request.treasuryFee = uint64(1 * 1e8); // 1 BTC
```
Comment (nkuba) [inline solidity/contracts/test/TestTBTCDepositorProxy.sol:83]: Can we calculate it based on the funding deposit amount:

```suggestion
        request.treasuryFee = _depositTreasuryFeeDivisor > 0
            ? fundingOutputAmount / _depositTreasuryFeeDivisor
            : 0;
```

And make the divisor configurable:
```solidity
    function setDepositTreasuryFeeDivisor(uint64 value) external {
        _depositTreasuryFeeDivisor = value;
    }
```

File: solidity/contracts/integrator/TBTCDepositorProxy.sol [hunk 13]
```diff
@@ -0,0 +1,309 @@
...
+        // revealed deposits stay there forever. The transaction will revert
+        // if the deposit has already been revealed so, there is no need to do
+        // an explicit check here.
+        bridge.revealDepositWithExtraData(fundingTx, reveal, extraData);
+
+        return depositKey;
+    }
+
+    /// @notice Finalizes a deposit by calculating the amount of TBTC minted
+    ///         for the deposit and calling the `onDepositFinalized` callback
```
Comment (nkuba) [inline solidity/contracts/integrator/TBTCDepositorProxy.sol:167]: The callback part is no longer valid.

File: solidity/contracts/integrator/TBTCDepositorProxy.sol [hunk 14]
```diff
@@ -0,0 +1,309 @@
...
+///          }
+///
+///          function finalizeProcess(uint256 depositKey) external {
+///              (uint256 tbtcAmount, bytes32 extraData) = _finalizeDeposit(depositKey);
+///
+///              // Do something with the minted TBTC using context
+///              // embedded in the extraData.
+///          }
+///      }
+abstract contract TBTCDepositorProxy {
```
Comment (pdyraga) [inline solidity/contracts/integrator/TBTCDepositorProxy.sol:73]: Why not call it just `TBTCDepositor` or `AbstractTBTCDepositor`? The proxy suffix could suggest it is some sort of an upgradeable proxy contract.

File: solidity/contracts/integrator/TBTCDepositorProxy.sol [hunk 15]
```diff
@@ -0,0 +1,309 @@
...
+    ///        deposit must be either swept or optimistically minted.
+    /// @dev IMPORTANT NOTE: The tbtcAmount returned by this function is an
+    ///      approximation. See documentation of the `calculateTbtcAmount`
+    ///      responsible for calculating this value for more details.
+    // slither-disable-next-line dead-code
+    function _finalizeDeposit(uint256 depositKey)
+        internal
+        returns (uint256 tbtcAmount, bytes32 extraData)
+    {
+        require(pendingDeposits[depositKey], "Deposit not initialized");
```
Comment (pdyraga) [inline solidity/contracts/integrator/TBTCDepositorProxy.sol:185]: Do we need `pendingDeposits` mapping? I think it should be enough to check whether [the `deposit.depositor` value](https://github.com/keep-network/tbtc-v2/blob/609b6fedc9444688092f310c38c0e03758bfdbfb/solidity/contracts/bridge/Deposit.sol#L102) is non-zero or set to `address(this)`. We retrieve `IBridgeTypes.DepositRequest` anyway. Also, the implementation of the `TBTCDepositorProxy` is already tightly coupled with the implementation of `Bridge` and `TBTCVault`. We could save some gas especially since those functions will be called by a relayer bot and not by the users.

EDIT: No, the current version is correct. We do not want `_finalizeDeposit` to be called twice for the same deposit.

File: solidity/contracts/integrator/TBTCDepositorProxy.sol [hunk 16]
```diff
@@ -0,0 +1,309 @@
...
+    /// @param depositTreasuryFeeSat Deposit treasury fee in satoshi (1e8 precision).
+    ///        This is an accurate value of the treasury fee that was actually
+    ///        cut upon minting.
+    /// @return tbtcAmount Approximate amount of TBTC minted for the deposit.
+    /// @dev IMPORTANT NOTE: The tbtcAmount returned by this function may
+    ///      not correspond to the actual amount of TBTC minted for the deposit.
+    ///      Although the treasury fee cut upon minting is known precisely,
+    ///      this is not the case for the optimistic minting fee and the Bitcoin
+    ///      transaction fee. To overcome that problem, this function just takes
+    ///      the current maximum values of both fees, at the moment of deposit
```
Comment (pdyraga) [inline solidity/contracts/integrator/TBTCDepositorProxy.sol:233]:
```suggestion
    ///      the current maximum allowed values of both fees, at the moment of deposit
```

File: solidity/contracts/integrator/TBTCDepositorProxy.sol (hunk 10 above)
Comment (pdyraga) [reply solidity/contracts/integrator/TBTCDepositorProxy.sol:183]: I agree with @nkuba, this could be a common use case and we can avoid reading from the storage one more time if we return the value here.

**Review by pdyraga — COMMENTED** [review]
I did a high-level review. The contract looks good to me with some potential minor tweaks already noted. I was not reviewing tests.

File: solidity/contracts/integrator/TBTCDepositorProxy.sol (hunk 10 above)
Comment (nkuba) [reply solidity/contracts/integrator/TBTCDepositorProxy.sol:183]: I proposed a change in https://github.com/keep-network/tbtc-v2/pull/779.

File: solidity/contracts/test/TestTBTCDepositorProxy.sol (hunk 11 above)
Comment (nkuba) [reply solidity/contracts/test/TestTBTCDepositorProxy.sol:79]: I proposed a change in https://github.com/keep-network/tbtc-v2/pull/779.

File: solidity/contracts/test/TestTBTCDepositorProxy.sol (hunk 12 above)
Comment (nkuba) [reply solidity/contracts/test/TestTBTCDepositorProxy.sol:83]: I proposed a change in https://github.com/keep-network/tbtc-v2/pull/779.

File: solidity/contracts/test/TestTBTCDepositorProxy.sol (hunk 9 above)
Comment (lukasz-zimnoch) [reply solidity/contracts/test/TestTBTCDepositorProxy.sol:96]: Leaving this for the integrator will give them more flexibility while writing tests. The second argument is that this abstract contract is unaware of the TBTC token and only pure numbers matter here.

File: solidity/contracts/integrator/TBTCDepositorProxy.sol (hunk 13 above)
Comment (lukasz-zimnoch) [reply solidity/contracts/integrator/TBTCDepositorProxy.sol]: Good catch. See db599050

File: solidity/contracts/integrator/TBTCDepositorProxy.sol (hunk 15 above)
Comment (lukasz-zimnoch) [reply solidity/contracts/integrator/TBTCDepositorProxy.sol:184]: > EDIT: No, the current version is correct. We do not want _finalizeDeposit to be called twice for the same deposit.

Exactly. Moreover, we don't want to push this responsibility to the child contract.

File: solidity/contracts/integrator/TBTCDepositorProxy.sol (hunk 16 above)
Comment (lukasz-zimnoch) [reply solidity/contracts/integrator/TBTCDepositorProxy.sol]: See a6f4a927

File: solidity/contracts/integrator/TBTCDepositorProxy.sol (hunk 14 above)
Comment (lukasz-zimnoch) [reply solidity/contracts/integrator/TBTCDepositorProxy.sol:73]: I wanted to avoid pure "depositor" to not mix it with the concept of "depositor" already present in the `Bridge`. I don't have strong preferences here though. `AbstractTBTCDepositor` is completely fine.

Comment (lukasz-zimnoch) [reply solidity/contracts/integrator/TBTCDepositorProxy.sol]: See 68d02e86

**Review by nkuba — APPROVED** [review]
